[Form] allow `entry_options` option of CollectionType to be dynamic

HeahDude opened the pull request:
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | n/a |
| License | MIT |
| Doc PR | symfony/symfony-docs#6319 |
- Deprecate usage of `prototype_data` option in favor of `prototype_options`.
- The `ResizeFormListener` now passes each entry of the collection to the `entry_options` option if it is a callable.

HeahDude committed:
[Form] deprecate 'prototype_data' option in favor of 'prototype_options' in CollectionType

HeahDude committed:
[Form] allow dynamic 'entry_options' in CollectionType

HeahDude commented:
To go further with this implementation, I think `CollectionList` and `ChoiceList` should use a common underlying `ListInterface` which would represent an editable list of form type entries, this should simplify things to deal with #9310.

webmozart commented on src/Symfony/Component/Form/Tests/Extension/Core/Type/CollectionTypeTest.php:
```diff
@@ -315,4 +343,172 @@ public function testPrototypeSetNotRequired()
+            'bar',
+        );
+
+        $attr = array(
```
I would name this `$expectedAttr` and move it below the `create()` call to make the intent of the test clear. Usually, tests should be structured into the following sections (which may in part be missing):

1. Fixture setup
2. Execution of the test
3. Test verification
4. Cleanup

The `create()` statement belongs to 2, while `$attr` (`$expectedAttr`) is already part of the test verification. Hence I'd move it below.

HeahDude replied:
Understood, thanks, I'm fixing it.

webmozart commented:
Hey @HeahDude, thanks for working on this :) Before you go any further, I'd finish the PR as is and start any new work in a new PR. This looks good so far.

HeahDude commented:
It's as easy as update title and description and open a new PR for next commits ;)

Thank you very much for your encouraging feedback :)

iltar commented:
What does this PR actually solve?

HeahDude commented:
The `CollectionType` can nest any other form types including itself.
Some form types has options that might be useful to set depending on the entry that populates its data.

Example:
```php
$builder->add('categories', CollectionType::class, array(
    'entry_type' => ChoiceType::class,
    'entry_options' => function ($category) {
        if (null !== $category) {
            return array(
                'choices' => $category->getTags(),
            );
        }

        return array();
    },
    ...

// or
$builder->add('metadata', CollectionType::class, array(
    'entry_type' => HiddenType::class,
    'entry_options' => function ($metadata) {
        $options = array();

        if (null !== $metadata) {
            $options['attr'] = array(
                'data-id' => $metadata->getId(),
                // ...
            );
        }

        return $options;
    },
// ...

But this will be even more powerful when `CollectionType` will be polymorphic as the callable will get the form type corresponding to the entry as second argument (PR coming soon).

stof commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,78 @@ public function finishView(FormView $view, FormInterface $form, array $options)
-        $entryOptionsNormalizer = function (Options $options, $value) {
-            $value['block_name'] = 'entry';
+        $entryOptionsNormalizer = function (Options $options, $entryOptions) {
+            if (is_array($entryOptions)) {
```
this is broken in case the callable is an array (something like ``array($this, 'buildEntryOptions')``)

HeahDude replied:
@stof, thanks I'll fix it and add a test for this.

stof replied:
btw, please add a test using such a callable to prevent regressions in its support

HeahDude replied:
ok looks like I got your tother comments with delay :) woking on it.

iltar commented:
I see, that indeed looks like a nice feature to have!

stof commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,78 @@ public function finishView(FormView $view, FormInterface $form, array $options)
+        };
+
+        $prototypeOptionsNormalizer = function (Options $options, $value) {
+            $entryOptions = is_array($options['entry_options'])
```
same issue here

stof commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,78 @@ public function finishView(FormView $view, FormInterface $form, array $options)
+                    $options['entry_options'],
+                    // For BC
+                    $options['prototype_data'] ?: (isset($value['data'])
+                        ? $value['data'] : null)
```
this should stay on the previous line

stof commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,78 @@ public function finishView(FormView $view, FormInterface $form, array $options)
+
+        $prototypeOptionsNormalizer = function (Options $options, $value) {
+            $entryOptions = is_array($options['entry_options'])
+                ? $options['entry_options']
```
Please use a ``if`` rather than a ternary. Nesting 3 ternary operators is totally unreadable

HeahDude commented:
Btw
```php
$builder->add('metadata', CollectionType::class, array(
    'entry_type' => HiddenType::class,
    'entry_options' => function ($metadata) {
        $options = array();

        if (null !== $metadata) {
            $options['attr'] = array(
                'data-id' => $metadata->getId(),
                // ...
            );
        }

        return $options;
    },
```

Should avoid the need of 'hidden' as `widget` option in `ChoiceType`, ref #6602

HeahDude commented:
Ok @stof, comments addressed :)

HeahDude commented:
Ok, since it would be too much conflict to harden options in 2.x and 3.0 is not LTS, I've squashed `prototype_options` handling. So there is now 2 commits for 2 bullets.

Also I've changed the scope of `ResizeFormListener::getOptionsForEntry()` from `private` to `protected` to avoid a BC break.

Doc PR adressed, ref symfony/symfony-docs#6319.

Green tests (failures still related to PHP 7 build and bootstrap form layout tests in Twig Bridge).

This one looks finished to me, needs final review. (and maybe a feature label :)

webmozart commented on src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:
```diff
@@ -125,7 +128,7 @@ public function preSubmit(FormEvent $event)
                     $form->add($name, $this->type, array_replace(array(
                         'property_path' => '['.$name.']',
-                    ), $this->options));
+                    ), $this->getOptionsForEntry($value)));
```
I just realize that there is a big issue with this solution. Above, in `preSetData()`, the method `getOptionsForEntry()` receives the data in the model format. For example, if we have a collection of `Tag` instances, a `Tag` instance would be passed.

Here, however, we pass the submitted data - a string (the tag name) or an array (multiple tag properties) - which are yet to be transformed back into a `Tag` instance. Consequently, the callback needs to be aware of either format and dynamically handle both. That won't be very intuitive nor easy to use in practice.

HeahDude replied:
Right, I did not see that. So I need to add a test on submit and I think I can handle this.

webmozart replied:
How would you handle this?

HeahDude replied:
It's funny because I was thinking about it last night and I wondered how was the `prototype_data` option handled though, and I realised while remembering your comment that you can only define the view data of a prototype. The `CollectionType` doc reference should be clear about it, especially if you need to transform value.

And just this morning, I see this ticket #17992...

Since there is no problem as long as the underlying data is a string value, it may not be working for an object underlying data.

I see an easy solution but it's not "pretty", it should be mentioned in the docs that the `entry_options` option as callable must test the string type :

```php
'entry_options' => function ($entry) {
    $options = array(
        // set default
    );

    if (is_string($entry) {
        // $entry <=> $viewData
        // or $entry <=> $modelData <=> $viewData
        $options['global_dynamic'] = // use $entry as string
    }

    // Else entry is an instanceof `data_class` or null
    $options['model_dynamic'] = // use $entry as model data or null

    return $options;
},
```

But it would be better to solve this in a feature way, by adding for example an `add_entry` option (which may have its counterpart in a future `allow_add`'able `ChoiceType` as `add_choice`).

This option would be a callable getting passed the new view data as only argument and should return a new model data. Almost like `empty_data` but this last one is only called for an empty view data and takes the form as first argument.

Then we would be able to use `add_entry` call on the new view data before passing it to `getOptionsForEntry`.

Another solution is to try to fix this while refactoring `CollectionType` more globally thanks to a `CollectionList` support, since I'm already investigating this way.
I think this would allow to solve many old issues with this type, with at first dynamic indexing, and polymorphism.

What do you think ?

HeahDude replied:
Of courses this is more a DX concern as one could use `DataMapperInterface` for that, but I'm not sure it's working while the new fields are added on PRE_SUBMIT event.

webmozart replied:
`prototype_data` accepts data in the model format. So you could actually pass an object, which would be converted to the right view format by the prototype field.

The problem we have here is that the field which is responsible for converting the view data back to the model data has not been created yet. At the same time, we cannot create the field yet since we still need to determine the options for the field creation by inspecting the model data. We have a circular dependency.

HeahDude replied:
But an `add_entry` option would be part of the parent scope (collection type), so could be accessible on in the `ResizeFormListener` which have access to the view data of the field when it adds it, right ?

webmozart replied:
The view data is not the problem :) Think about when we need to call that closure:

* During `setData()`: Here we want to create the fields for a collection in model format
* During `submit()`: Here we want to create the fields for an array in view format

It is technically impossible (it seems to me) to solve this problem with one closure that accepts only either the model format or the view format. Dealing with both formats in the closure, on the other hand, is not really usable.

HeahDude replied:
I'm going to add a test for it anyway, as for the behavior of `prototype_data` and `empty_data` on submit to avoid regressions such as #17992.

I keep investigating, I think this feature can be useful and doable even if it needs a more global refactoring of `CollectionType`.

It may not be finished for 3.1 though, but I'll try to word on that this week-end.

stof replied:
@HeahDude even with a refactoring, we will still need to add elements based on 2 different formats, as one case wants to add fields when knowing the model data and the other case when knowing the view data (and this cannot change with any refactoring, as it is part of the basic about how a collection type should work).

Btw, one of your previous comment assumes that the view data is a string. this assumption is wrong. you can perfectly have collections of objects (and then a compound form type as the entry type)

HeahDude replied:
I agree with don't misunderstood me, I would assume default view data is string when not null, or when compound an array or object.

In fact, I mentioned that `ChoiceType` is particular in this sense since it is always compound (whether it can be a select input with no children) and will always have a string view data.
That what I would like to ensure decoupling the data transformers from the list mapper in #18180.

HeahDude replied:
In fact I think expanded `ChoiceType` is not far from the `CollectionType` and I was thinking about decoupling the form list creation with a factory interface as #14050 did for `ChoiceType` to allow dealing with polymorphic and custom indexed/named collection.

HeahDude replied:
and dynamic entry options.

HeahDude commented:
Status: Needs work

HeahDude commented:
@webmozart What do you think about merging c969206 in 3.1, since `prototype_data` has never been added to the doc anyway and it could be useful to define specific `attr` to prototype.

In that case I'll try to keep working on `CollectionType` in a later PR.

HeahDude commented:
Status: Needs Review

for c969206

trousers commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,80 @@ public function finishView(FormView $view, FormInterface $form, array $options)
+            return $entryOptions;
+        };
+
+        $prototypeOptionsNormalizer = function (Options $options, $value) {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid excessively long variable names like $prototypeOptionsNormalizer. Keep variable name length under 20.](https://www.codacy.com/app/git_9/symfony/file/2332080378/issues/source?bid=3207821&fileBranchId=3207880#l98)

trousers commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,80 @@ public function finishView(FormView $view, FormInterface $form, array $options)
     {
-        $entryOptionsNormalizer = function (Options $options, $value) {
-            $value['block_name'] = 'entry';
+        $entryOptionsNormalizer = function (Options $options, $entryOptions) {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [Avoid excessively long variable names like $entryOptionsNormalizer. Keep variable name length under 20.](https://www.codacy.com/app/git_9/symfony/file/2332080378/issues/source?bid=3207821&fileBranchId=3207880#l76)

trousers commented on src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php:
```diff
@@ -81,24 +73,80 @@ public function finishView(FormView $view, FormInterface $form, array $options)
+                    // BC, 'prototype_data' will be removed in 4.0.
+                    $options['prototype_data'] ?: (isset($value['data']) ? $value['data'] : null)
+                );
+            } else {
```
![Codacy](https://www.codacy.com/assets/images/favicon.png) Issue found: [The method configureOptions uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/git_9/symfony/file/2332080378/issues/source?bid=3207821&fileBranchId=3207880#l106)

xabbuh commented on src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:
```diff
@@ -29,7 +29,9 @@ class ResizeFormListener implements EventSubscriberInterface
 
     /**
-     * @var array
+     * A callable get passed each entry as only argument.
```
[...] gets passed [...] as the only [...]

HeahDude commented:
Status: Needs Work

fabpot commented:
@HeahDude What's the status of this PR?

HeahDude commented:
Not ready for 3.2 sadly. I'll keep working on that for sure!

fabpot commented:
@HeahDude  Last call for 3.3 :)

HeahDude commented:
I think the one will be easier to implement in symfony 4.x :). Closing for now.
